File: python/neutron/nucleus/timeseries.py
"""TimeSeries model — wraps Nucleus TS_* SQL functions."""

from __future__ import annotations

import json
from typing import cast

from datetime import datetime, timedelta, timezone

from pydantic import BaseModel

from neutron.nucleus._exec import Executor, require_nucleus
from neutron.nucleus.client import Features


class TimeSeriesPoint(BaseModel):
    timestamp: datetime
    value: float
    tags: dict[str, str] = {}


class TimeSeriesModel:
    """Time-series data operations over Nucleus.

    Usage::

        await db.timeseries.write("cpu.usage", [
            TimeSeriesPoint(timestamp=now, value=72.5),
        ])
        points = await db.timeseries.query("cpu.usage", start, end)
    """

    def __init__(self, executor: Executor, features: Features) -> None:
        self._exec = executor
        self._features = features

    def _require(self) -> None:
        require_nucleus(self._features, "TimeSeries")

    async def write(
        self, measurement: str, points: list[TimeSeriesPoint]
    ) -> None:
        """Insert time-series data points."""
        self._require()
        for point in points:
            ts_ms = int(point.timestamp.timestamp() * 1000)
            await self._exec.fetchval(
                "SELECT TS_INSERT($1, $2, $3)", measurement, ts_ms, point.value
            )

    async def last(self, measurement: str) -> float | None:
        """Get the latest value for a series."""
        self._require()
        return cast("float | None", await self._exec.fetchval("SELECT TS_LAST($1)", measurement))

    async def count(self, measurement: str) -> int:
        """Count data points in a series."""
        self._require()
        return cast("int", await self._exec.fetchval("SELECT TS_COUNT($1)", measurement))

    async def query(
        self,
        measurement: str,
        start: datetime,
        end: datetime,
        *,
        tags: dict[str, str] | None = None,
    ) -> list[TimeSeriesPoint]:
        """Return the raw data points stored in ``[start, end]``.

        This used to synthesize the answer from ``buckets`` (default 60)
        ``TS_RANGE_AVG`` calls, because raw point retrieval had no SQL surface.
        That was sixty round trips to read points the engine already held, and
        it was lossy: any bucket containing more than one point returned their
        average as a single fabricated point at the bucket boundary, so the
        timestamps did not correspond to stored data.

        ``TS_RANGE`` now returns the points themselves, and every SDK uses it,
        which is what closes the three-different-answers gap this method was on
        the wrong side of. Use ``aggregate`` when bucketed averages are what is
        wanted — that is a different question and has its own method.
        """
        self._require()
        start_ms = int(start.timestamp() * 1000)
        end_ms = int(end.timestamp() * 1000)
        if end_ms <= start_ms:
            return []

        raw = await self._exec.fetchval(
            "SELECT TS_RANGE($1, $2, $3)", measurement, start_ms, end_ms
        )
        if not raw:
            return []
        return [
            TimeSeriesPoint(
                timestamp=datetime.fromtimestamp(item["t"] / 1000, tz=timezone.utc),
                value=float(item["v"]),
            )
            for item in json.loads(raw)
        ]

    async def aggregate(
        self,
        measurement: str,
        start: datetime,
        end: datetime,
        window: timedelta,
        fn: str = "avg",
    ) -> list[TimeSeriesPoint]:
        """Aggregate data points with time bucketing.

        Returns one data point per ``window``-sized bucket across the range.

        ``fn`` is ``avg`` or ``count``. The docstring used to promise ``sum``,
        ``min``, ``max``, ``first`` and ``last`` as well; the engine ships
        TS_RANGE_AVG and TS_RANGE_COUNT and nothing else, so those five raised
        ValueError from the map below. Documenting what exists rather than what
        was planned.
        """
        self._require()
        start_ms = int(start.timestamp() * 1000)
        end_ms = int(end.timestamp() * 1000)
        window_ms = int(window.total_seconds() * 1000)
        if window_ms <= 0 or end_ms <= start_ms:
            return []

        # Map fn to the correct Nucleus TS_RANGE_* SQL function
        fn_map = {
            "avg": "TS_RANGE_AVG",
            "count": "TS_RANGE_COUNT",
        }
        sql_fn = fn_map.get(fn.lower())
        if sql_fn is None:
            raise ValueError(
                f"Unsupported aggregation function: {fn}. "
                f"Supported: {', '.join(fn_map)}"
            )

        # Align start to a TIME_BUCKET boundary.
        #
        # This passed an interval NAME here — "minute"/"hour"/"day"/"week" —
        # chosen by bucketing the window size. The engine's TIME_BUCKET takes
        # (bucket_millis, ts), both INT8, so every call raised on the type and
        # aggregate() had never once worked. Two bugs in one: even had the
        # string been accepted, aligning a 5-minute window to an hour boundary
        # would have produced buckets that do not line up with the window the
        # caller asked for. Align to window_ms, which is what "align to the
        # bucket size" means.
        aligned = await self._exec.fetchval(
            "SELECT TIME_BUCKET($1, $2)", window_ms, start_ms
        )
        bucket_start = int(aligned) if aligned is not None else start_ms

        points: list[TimeSeriesPoint] = []
        while bucket_start < end_ms:
            bucket_end = bucket_start + window_ms
            effective_end = min(bucket_end, end_ms)
            result = await self._exec.fetchval(
                f"SELECT {sql_fn}($1, $2, $3)",
                measurement,
                bucket_start,
                effective_end,
            )
            if result is not None:
                ts = datetime.fromtimestamp(bucket_start / 1000, tz=timezone.utc)
                points.append(TimeSeriesPoint(timestamp=ts, value=float(result)))
            bucket_start = bucket_end

        return points

    async def retention(self, max_age_ms: int) -> bool:
        """Set the global retention policy for all time series.

        The engine's TS_RETENTION takes a single ``max_age_ms`` argument and
        applies it globally — per-series retention is not supported.
        """
        self._require()
        result = await self._exec.fetchval("SELECT TS_RETENTION($1)", max_age_ms)
        return result == "OK"

File: elixir/lib/nucleus/models/timeseries.ex
defmodule Nucleus.Models.TimeSeries do
  @moduledoc """
  Time-series model — TS_INSERT, TS_LAST, TS_COUNT, TS_RANGE, TIME_BUCKET.

  ## Example

      alias Nucleus.Models.TimeSeries

      TimeSeries.insert(client, "cpu_usage", System.system_time(:millisecond), 72.5)
      {:ok, 72.5} = TimeSeries.last(client, "cpu_usage")
      {:ok, avg} = TimeSeries.range_avg(client, "cpu_usage", start_ms, end_ms)
  """

  @type client :: Nucleus.Client.t()

  @doc "Inserts a time-series data point."
  @spec insert(client(), String.t(), integer(), number()) :: :ok | {:error, term()}
  def insert(client, series, timestamp_ms, value) do
    with :ok <- Nucleus.Client.require_nucleus(client, "TimeSeries.insert") do
      case Nucleus.Client.query(client, "SELECT TS_INSERT($1, $2, $3)", [
             series,
             timestamp_ms,
             value
           ]) do
        {:ok, _} -> :ok
        {:error, _} = error -> error
      end
    end
  end

  @doc "Returns the last value in a series."
  @spec last(client(), String.t()) :: {:ok, float() | nil} | {:error, term()}
  def last(client, series) do
    with :ok <- Nucleus.Client.require_nucleus(client, "TimeSeries.last") do
      case Nucleus.Client.query(client, "SELECT TS_LAST($1)", [series]) do
        {:ok, %{rows: [[val]]}} -> {:ok, val}
        {:ok, %{rows: []}} -> {:ok, nil}
        {:error, _} = error -> error
      end
    end
  end

  @doc "Returns the count of data points in a series."
  @spec count(client(), String.t()) :: {:ok, integer()} | {:error, term()}
  def count(client, series) do
    with :ok <- Nucleus.Client.require_nucleus(client, "TimeSeries.count") do
      case Nucleus.Client.query(client, "SELECT TS_COUNT($1)", [series]) do
        {:ok, %{rows: [[val]]}} -> {:ok, val}
        {:error, _} = error -> error
      end
    end
  end

  @doc "Returns the count of data points in a time range."
  @spec range_count(client(), String.t(), integer(), integer()) ::
          {:ok, integer()} | {:error, term()}
  def range_count(client, series, start_ms, end_ms) do
    with :ok <- Nucleus.Client.require_nucleus(client, "TimeSeries.range_count") do
      case Nucleus.Client.query(client, "SELECT TS_RANGE_COUNT($1, $2, $3)", [
             series,
             start_ms,
             end_ms
           ]) do
        {:ok, %{rows: [[val]]}} -> {:ok, val}
        {:error, _} = error -> error
      end
    end
  end

  @doc "Returns the average value in a time range."
  @spec range_avg(client(), String.t(), integer(), integer()) ::
          {:ok, float() | nil} | {:error, term()}
  def range_avg(client, series, start_ms, end_ms) do
    with :ok <- Nucleus.Client.require_nucleus(client, "TimeSeries.range_avg") do
      case Nucleus.Client.query(client, "SELECT TS_RANGE_AVG($1, $2, $3)", [
             series,
             start_ms,
             end_ms
           ]) do
        {:ok, %{rows: [[val]]}} -> {:ok, val}
        {:ok, %{rows: []}} -> {:ok, nil}
        {:error, _} = error -> error
      end
    end
  end

  @doc "Sets the global retention policy (auto-delete points older than `max_age_ms`)."
  @spec retention(client(), integer()) :: :ok | {:error, term()}
  def retention(client, max_age_ms) do
    with :ok <- Nucleus.Client.require_nucleus(client, "TimeSeries.retention") do
      case Nucleus.Client.query(client, "SELECT TS_RETENTION($1)", [max_age_ms]) do
        {:ok, _} -> :ok
        {:error, _} = error -> error
      end
    end
  end

  @doc """
  Aggregates a series into fixed windows across a range.

  Returns one `{bucket_start_ms, value}` per `window_ms`-sized bucket between
  `start_ms` and `end_ms`, skipping buckets with no data.

  `fn` is `:avg` or `:count`. The engine ships `TS_RANGE_AVG` and
  `TS_RANGE_COUNT` and nothing else — sum/min/max/first/last do not exist, so
  they are rejected here rather than silently returning an average.

  Two traps this inherits from the Python implementation, both of which cost a
  release there:

    * `TIME_BUCKET` takes `(bucket_millis, ts)`, both integers. Passing an
      interval NAME — "minute", "hour" — raises on the type, and that bug meant
      Python's `aggregate` had never once worked.
    * Alignment is to `window_ms`, not to a calendar unit. Aligning a
      five-minute window to an hour boundary produces buckets that do not line
      up with the window the caller asked for, which is a wrong answer rather
      than an error.
  """
  @spec aggregate(client(), String.t(), integer(), integer(), integer(), :avg | :count) ::
          {:ok, [{integer(), float()}]} | {:error, term()}
  def aggregate(client, series, start_ms, end_ms, window_ms, fun \\ :avg) do
    with :ok <- Nucleus.Client.require_nucleus(client, "TimeSeries.aggregate"),
         {:ok, sql_fn} <- aggregate_fn(fun) do
      cond do
        window_ms <= 0 -> {:ok, []}
        end_ms <= start_ms -> {:ok, []}
        true -> do_aggregate(client, series, start_ms, end_ms, window_ms, sql_fn)
      end
    end
  end

  defp aggregate_fn(:avg), do: {:ok, "TS_RANGE_AVG"}
  defp aggregate_fn(:count), do: {:ok, "TS_RANGE_COUNT"}

  defp aggregate_fn(other),
    do: {:error, {:unsupported_aggregate, other, [:avg, :count]}}

  defp do_aggregate(client, series, start_ms, end_ms, window_ms, sql_fn) do
    with {:ok, aligned} <- time_bucket(client, window_ms, start_ms) do
      bucket_start = aligned || start_ms

      bucket_start
      |> Stream.iterate(&(&1 + window_ms))
      |> Stream.take_while(&(&1 < end_ms))
      |> Enum.reduce_while({:ok, []}, fn bstart, {:ok, acc} ->
        effective_end = min(bstart + window_ms, end_ms)

        case Nucleus.Client.query(client, "SELECT #{sql_fn}($1, $2, $3)", [
               series,
               bstart,
               effective_end
             ]) do
          {:ok, %{rows: [[nil]]}} -> {:cont, {:ok, acc}}
          {:ok, %{rows: [[v]]}} -> {:cont, {:ok, [{bstart, v / 1} | acc]}}
          {:ok, %{rows: []}} -> {:cont, {:ok, acc}}
          {:error, _} = error -> {:halt, error}
        end
      end)
      |> case do
        {:ok, acc} -> {:ok, Enum.reverse(acc)}
        other -> other
      end
    end
  end

  @doc """
  Truncates a timestamp (ms) down to its bucket of `bucket_ms` milliseconds.
  """
  @spec time_bucket(client(), integer(), integer()) :: {:ok, integer()} | {:error, term()}
  def time_bucket(client, bucket_ms, timestamp) do
    with :ok <- Nucleus.Client.require_nucleus(client, "TimeSeries.time_bucket") do
      case Nucleus.Client.query(client, "SELECT TIME_BUCKET($1, $2)", [bucket_ms, timestamp]) do
        {:ok, %{rows: [[val]]}} -> {:ok, val}
        {:error, _} = error -> error
      end
    end
  end
end
